x-pack/plugins/security/server/saved_objects/saved_objects_security_extension.ts

```diff
@@ -139,33 +229,114 @@ export class SavedObjectsSecurityExtension implements ISavedObjectsSecurityExten
+      for (const obj of objects) {
+        this.addAuditEvent({
+          action,
+          savedObject: { type: obj.type, id: obj.id },
+          error,
+          // By default, if authorization was a success the outcome is 'unknown' because the operation has not occurred yet
+          // The GET action is one of the few exceptions to this, and hence it passes true to useSuccessOutcome
+          ...(!error && { outcome: useSuccessOutcome ? 'success' : 'unknown' }),
+          addToSpaces,
+          deleteFromSpaces,
+        });
+      }
+    } else {
+      this.addAuditEvent({
+        action,
+        error,
+        // outcome: error ? 'failure' : useSuccessOutcome ? 'success' : 'unknown',
+        ...(!error && { outcome: useSuccessOutcome ? 'success' : 'unknown' }),
+        addToSpaces,
+        deleteFromSpaces,
+      });
+    }
+  }
+
   enforceAuthorization<A extends string>(params: EnforceAuthorizationParams<A>): void {
-    const { typesAndSpaces, action, typeMap, auditCallback } = params;
+    const { typesAndSpaces, action, typeMap, auditOptions } = params;
+    const {
+      objects: auditObjects,
+      bypassOnSuccess: bypassAuditOnSuccess,
+      bypassOnFailure: bypassAuditOnFailure,
+      useSuccessOutcome,
+      addToSpaces,
+      deleteFromSpaces,
+    } = auditOptions ?? {};
+
+    const { authzAction, auditAction } = this.decodeSecurityAction(action);
+
     const unauthorizedTypes = new Set<string>();
-    for (const [type, spaces] of typesAndSpaces) {
-      const spacesArray = [...spaces];
-      if (!isAuthorizedInAllSpaces(type, action, spacesArray, typeMap)) {
-        unauthorizedTypes.add(type);
+
+    if (authzAction) {
+      for (const [type, spaces] of typesAndSpaces) {
+        const spacesArray = [...spaces];
+        if (!isAuthorizedInAllSpaces(type, authzAction as A, spacesArray, typeMap)) {
+          unauthorizedTypes.add(type);
+        }
       }
     }
 
     if (unauthorizedTypes.size > 0) {
       const targetTypes = [...unauthorizedTypes].sort().join(',');
-      const msg = `Unable to ${action} ${targetTypes}`;
+      const msg = `Unable to ${authzAction} ${targetTypes}`;
       const error = this.errors.decorateForbiddenError(new Error(msg));
-      auditCallback?.(error);
+      if (auditAction && !bypassAuditOnFailure) {
+        this.auditHelper(
+          auditAction,
+          auditObjects,
+          useSuccessOutcome,
+          addToSpaces,
+          deleteFromSpaces,
+          error
+        );
+      }
       throw error;
     }
-    auditCallback?.();
+
+    if (auditAction && !bypassAuditOnSuccess) {
+      this.auditHelper(auditAction, auditObjects, useSuccessOutcome);
+    }
   }
 
   async performAuthorization<A extends string>(
     params: PerformAuthorizationParams<A>
-  ): Promise<CheckAuthorizationResult<A>> {
-    const checkResult: CheckAuthorizationResult<A> = await this.checkAuthorization({
+  ): Promise<CheckAuthorizationResult<A> | undefined> {
```

Comment:
The "PerformAuthorization" method of the extension will likely be dissolved as it was originally implemented as a way to generically combine checkAuthorization and enforceAuthorization. If we move forward with implementing per-action authorization methods, we will not need this method.